data/make_readme.py

```diff
@@ -101,87 +89,120 @@ def validate_item(itm: Item, section_title: str) -> None:
         raise ValueError("\n".join(errors))
 
 
-for key, section in sections.items():
-    # Keep lang_names inside sections loop to refill language subsections for each new
-    # section. Used by both repos and Packages. Is a list for order and mutability.
-    lang_names = ["PyTorch", "TensorFlow", "JAX", "Julia", "Other"]
-
-    # sort first by language with order determined by lang_names (only applies to
-    # Package and repos sections), then by date
-    section["items"].sort(key=lambda x: x["date"], reverse=True)
-    if key in ("packages", "repos"):
-        section["items"].sort(key=lambda itm: lang_names.index(itm["lang"]))
-
-    # add item count after section title
-    section["markdown"] += f" <small>({len(section['items'])})</small>\n\n"
-
-    for itm in section["items"]:
-        if (lang := itm.get("lang")) in lang_names:
-            lang_names.remove(lang)
-            # print language subsection title if this is the first item with that lang
-            section["markdown"] += (
-                f'<br>\n\n### <img src="assets/{lang.lower()}.svg" alt="{lang}" '
-                f'height="20px"> &nbsp;{lang} {key.title()}\n\n'
+if __name__ == "__main__":
+    titles = {
+        "publications": "## 📝 Publications",
+        "applications": "## 🛠️ Applications",
+        "videos": "## 📺 Videos",
+        "packages": "## 📦 Packages",
+        "repos": "## 🧑‍💻 Repos",
+        "posts": "## 🌐 Blog Posts",
+    }
+
+    sections: dict[str, Section] = {
+        key: {"title": titles[key], "items": load_items(key), "markdown": ""}
+        for key in titles  # markdown is set below
+    }
+
+    seen_titles: set[tuple[str, str]] = set()
+    required_keys = {"title", "url", "date", "authors", "description"}
+    optional_keys = {
+        "lang",
+        "repo",
+        "docs",
+        "date_added",
+        "last_updated",
+    }
+    valid_languages = {"PyTorch", "TensorFlow", "JAX", "Julia", "Other"}
```

Comment:
just noticed these are no longer in the module scope. which means `validate_item` would fail if you import it and try to call it from another file. not that there are any plans to do that. hence probably best to remove `if __name__ == "__main__":` altogether and reduce indentation level. i would also move these back up since i think it helps readability to have variables used in `validate_item` defined above it

Reply:
Ah sure (I haven't really went through the script yet). But I personally prefer to keep the `if __name__ == "__main__":` such that it would be clearer what is actually executed (though indeed increases the indentation level).